RZ multimode geometry

dpgrote opened the pull request:
This implements the multimode RZ solver. It touches a fair number of files because of the change in the number of components of all of the arrays. Note that this needs the FDTD_RZ_multimode branch of picsar.

The big change is adding the extra components to all of the arrays. Each field now has 2*nmodes components (where nmodes is the number of modes). The factor of 2 is because each mode is complex, having a real and imaginary piece. The base Fortran routines for the solver use complex numbers (which makes the code cleaner). When those routines are called, the code transposes the data, mapping from a real array of size (nr,nz,2*nmodes) to complex array of size (nr,nz,nmodes), and maps back if needed.

I'll make some more changes to this PR. One is to add more comments in the source code. The other is modify the plasma injection to allow evenly spaced particles in theta. (DONE)

One other piece to do yet is handle the writing of the fields for the plot files. Now, as a temporary thing, the code is only writing out the real part of the first mode of the fields (see the routine PackPlotDataPtrs). This will be changed so that the full fields are written out (summing over the modes). Note that the raw fields will include all of the modes already. (DONE)

dpgrote commented:
The particle loading now creates particles evenly spaced in theta (when nmodes > 1).

The writing of the fields to the plot file is now updated. It writes out the total field in the usual arrays (Ex, Ey, Ez), and also writes out all of the modes with the real and imaginary parts in separate arrays, with names like Er0_real (real part of Er for mode 0) , Etheta1_imag (imaginary part of Etheta for mode 1).

There is a test case Examples/Tests/Langmuir/inputs.multi.rzmodes.rt and the analysis file langmuir_multi_rz_multimode_analysis.py. The results compare well to the analytic solution, but the difference it greater than the few percent threshold used for other tests. I don't know if there is a bug or if the test case is too sensitive.

RemiLehe commented on Source/Particles/PhysicalParticleContainer.cpp:
```diff
@@ -41,16 +41,19 @@ NumParticlesToAdd(const Box& overlap_box, const RealBox& overlap_realbox,
 	    Real y = 0;
-	    Real z = overlap_corner[1] + (iv[1] + r[1])*dx[1];
+#ifdef WARPX_RZ
+	    Real z = overlap_corner[1] + (iv[1] + point[2])*dx[1];
```
Could you add a comment explaining the difference between RZ and 2D here? Because of the indices (`1`, `2`), it is not entirely transparent what quantity corresponds to `r`, `z` or `theta`.

RemiLehe commented on Source/Particles/PhysicalParticleContainer.cpp:
```diff
@@ -448,16 +451,19 @@ PhysicalParticleContainer::AddPlasmaCPU (int lev, RealBox part_realbox)
                     Real y = 0;
-                    Real z = overlap_corner[1] + (iv[1] + r[1])*dx[1];
+#ifdef WARPX_RZ
+                    Real z = overlap_corner[1] + (iv[1] + point[2])*dx[1];
```
Same remark here.

RemiLehe commented on Source/Particles/PhysicalParticleContainer.cpp:
```diff
@@ -1133,6 +1156,7 @@ PhysicalParticleContainer::FieldGather (int lev,
                BL_TO_FORTRAN_ANYD(bxfab),
 	       BL_TO_FORTRAN_ANYD(byfab),
 	       BL_TO_FORTRAN_ANYD(bzfab),
+           &WarpX::nmodes,
```
Could you correct the indentation?

dpgrote commented:
@RemiLehe  I think I've addressed all of your comments.

Note that this PR will have conflicts with the EBpush_to_cpp branch PR. I would say that the EBpush_to_cpp PR is more important to merge first, specifically before the hackathon.
